### PR TITLE
feat(core): Added integrity framework for Authorize and Sync flow with connector as Stripe

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [x] New feature

## Description

With this PR, we are implementing connector integrity checks for the authorize and sync flow. After receiving a response from the connector, we will validate on our end whether the data sent to the connector matches the data received from their end.

If the validation fails we will mark the transaction as processing <a non terminal state> so that if the connector fixes the response in sync, the status will move to respective terminal state.

At api level the response will be 5xx with connector_transaction_id <if present> in response for merchant to reference in case if integrity failures for both sync and authorize.

## How did you test it?

Since its a blackbox scenario, we cannot test until and unless there is a error data from connectors end!

Local testing scenario

##### Payments Create

```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_ng5enE9FcquL9C0eYO1Jr0E6U8V8vbriDijpN4EgVpr1o31zWAMrVzkIt3q5ZvRh' \
--data-raw '
{
    "amount": 250,
    "currency": "USD",
    "confirm": true,
    "amount_to_capture": 250,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "StripeCustomer123",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Please Pay for your XYZ",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "off_session",
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "routing": {
        "type": "single",
        "data": "nmi"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "999"
        }
    }
}
'
```
Screenshot 2024-06-24 at 4 34 48 PM

#####Payments Sync

with force_sync as False
```shell
curl --location 'http://localhost:8080/payments/pay_x99f0Q4DO2drPrt3QZX0' \
--header 'Accept: application/json' \
--header 'api-key: dev_ng5enE9FcquL9C0eYO1Jr0E6U8V8vbriDijpN4EgVpr1o31zWAMrVzkIt3q5ZvRh'
```

response
```shell
{
    "payment_id": "pay_x99f0Q4DO2drPrt3QZX0",
    "merchant_id": "merchant_1718267057",
    "status": "processing",
    "amount": 250,
    "net_amount": 250,
    "amount_capturable": 250,
    "amount_received": 250,
    "connector": "stripe",
    "client_secret": "pay_x99f0Q4DO2drPrt3QZX0_secret_0A6rPzdRv1pvBLdZ61GA",
    "created": "2024-06-24T11:03:39.655Z",
    "currency": "USD",
    "customer_id": "StripeCustomer123",
    "customer": {
        "id": "StripeCustomer123",
        "name": "John Doe",
        "email": "abcdef123@gmail.com",
        "phone": "999999999",
        "phone_country_code": "+65"
    },
    "description": "Please Pay for your XYZ",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": null,
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "payment_checks": null,
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": "Integrity Check Failed! Value mismatched for fields currency",
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "pi_3PVAgTCuXDMKrYZe0L9s7DNy",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_rhbVuxclv1iy3M2pZkkI",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_sKlZxW37AYQNFv4Z4LGy",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-06-24T11:18:39.655Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "127.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-06-24T11:03:42.546Z",
    "charges": null,
    "frm_metadata": null
}
```
we should get error message as integrity check failed with status as processing.

with force_sync as true

```shell
curl --location 'http://localhost:8080/payments/pay_x99f0Q4DO2drPrt3QZX0?force_sync=true' \
--header 'Accept: application/json' \
--header 'api-key: dev_ng5enE9FcquL9C0eYO1Jr0E6U8V8vbriDijpN4EgVpr1o31zWAMrVzkIt3q5ZvRh'
```
Screenshot 2024-06-24 at 4 38 30 PM

#### Scenario[2] Where Connector send wrong currency in both authorize but fixes the same sync flow

#####Payments Create

```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_ng5enE9FcquL9C0eYO1Jr0E6U8V8vbriDijpN4EgVpr1o31zWAMrVzkIt3q5ZvRh' \
--data-raw '
{
    "amount": 250,
    "currency": "USD",
    "confirm": true,
    "amount_to_capture": 250,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "StripeCustomer123",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Please Pay for your XYZ",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "off_session",
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "routing": {
        "type": "single",
        "data": "nmi"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "999"
        }
    }  
}
'
```
Screenshot 2024-06-24 at 4 41 51 PM

##### Payments Sync
with force_sync true

```shell
curl --location 'http://localhost:8080/payments/pay_ExsGuWmhYUqDrkkAJnVg?force_sync=true' \
--header 'Accept: application/json' \
--header 'api-key: dev_ng5enE9FcquL9C0eYO1Jr0E6U8V8vbriDijpN4EgVpr1o31zWAMrVzkIt3q5ZvRh'
```

Response
```shell
{
    "payment_id": "pay_ExsGuWmhYUqDrkkAJnVg",
    "merchant_id": "merchant_1718267057",
    "status": "succeeded",
    "amount": 250,
    "net_amount": 250,
    "amount_capturable": 0,
    "amount_received": 250,
    "connector": "stripe",
    "client_secret": "pay_ExsGuWmhYUqDrkkAJnVg_secret_upTfAFa5sOSTcGTMKKIO",
    "created": "2024-06-24T11:11:45.150Z",
    "currency": "USD",
    "customer_id": "StripeCustomer123",
    "customer": {
        "id": "StripeCustomer123",
        "name": "John Doe",
        "email": "abcdef123@gmail.com",
        "phone": "999999999",
        "phone_country_code": "+65"
    },
    "description": "Please Pay for your XYZ",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": null,
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "payment_checks": {
                "address_line1_check": null,
                "address_postal_code_check": null,
                "cvc_check": "pass"
            },
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "pi_3PVAoICuXDMKrYZe1GvfbMYa",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pi_3PVAoICuXDMKrYZe1GvfbMYa",
    "payment_link": null,
    "profile_id": "pro_rhbVuxclv1iy3M2pZkkI",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_sKlZxW37AYQNFv4Z4LGy",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-06-24T11:26:45.150Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "127.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-06-24T11:12:40.350Z",
    "charges": null,
    "frm_metadata": null
}
```
Status should move to succeeded!

#### Scenario[3] Where connector sends correct data in both authorize and sync flow

#####Payments Create
```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_ng5enE9FcquL9C0eYO1Jr0E6U8V8vbriDijpN4EgVpr1o31zWAMrVzkIt3q5ZvRh' \
--data-raw '
{
    "amount": 250,
    "currency": "USD",
    "confirm": true,
    "amount_to_capture": 250,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "StripeCustomer123",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Please Pay for your XYZ",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "off_session",
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "routing": {
        "type": "single",
        "data": "nmi"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "999"
        }
    }
}
'
```

Response

```shell
{
    "payment_id": "pay_HD0mGNbh3sY4SDzEJsPv",
    "merchant_id": "merchant_1718267057",
    "status": "succeeded",
    "amount": 250,
    "net_amount": 250,
    "amount_capturable": 0,
    "amount_received": 250,
    "connector": "stripe",
    "client_secret": "pay_HD0mGNbh3sY4SDzEJsPv_secret_EpmbV93ry8SiEMGIjCDF",
    "created": "2024-06-24T11:16:26.868Z",
    "currency": "USD",
    "customer_id": "StripeCustomer123",
    "customer": {
        "id": "StripeCustomer123",
        "name": "John Doe",
        "email": "abcdef123@gmail.com",
        "phone": "999999999",
        "phone_country_code": "+65"
    },
    "description": "Please Pay for your XYZ",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": null,
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "payment_checks": {
                "address_line1_check": null,
                "address_postal_code_check": null,
                "cvc_check": "pass"
            },
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer123",
        "created_at": 1719227786,
        "expires": 1719231386,
        "secret": "epk_3f099e2eb5f74a1db265f81af40082b3"
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "pi_3PVAsqCuXDMKrYZe1FwwRbez",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pi_3PVAsqCuXDMKrYZe1FwwRbez",
    "payment_link": null,
    "profile_id": "pro_rhbVuxclv1iy3M2pZkkI",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_sKlZxW37AYQNFv4Z4LGy",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-06-24T11:31:26.868Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "127.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-06-24T11:16:29.856Z",
    "charges": null,
    "frm_metadata": null
}
```

#####Payments Sync

```shell
curl --location 'http://localhost:8080/payments/pay_HD0mGNbh3sY4SDzEJsPv?force_sync=true' \
--header 'Accept: application/json' \
--header 'api-key: dev_ng5enE9FcquL9C0eYO1Jr0E6U8V8vbriDijpN4EgVpr1o31zWAMrVzkIt3q5ZvRh'
```

Response

```shell
{
    "payment_id": "pay_HD0mGNbh3sY4SDzEJsPv",
    "merchant_id": "merchant_1718267057",
    "status": "succeeded",
    "amount": 250,
    "net_amount": 250,
    "amount_capturable": 0,
    "amount_received": 250,
    "connector": "stripe",
    "client_secret": "pay_HD0mGNbh3sY4SDzEJsPv_secret_EpmbV93ry8SiEMGIjCDF",
    "created": "2024-06-24T11:16:26.868Z",
    "currency": "USD",
    "customer_id": "StripeCustomer123",
    "customer": {
        "id": "StripeCustomer123",
        "name": "John Doe",
        "email": "abcdef123@gmail.com",
        "phone": "999999999",
        "phone_country_code": "+65"
    },
    "description": "Please Pay for your XYZ",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": null,
            "card_exp_month": "08",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "payment_checks": {
                "cvc_check": "pass",
                "address_line1_check": null,
                "address_postal_code_check": null
            },
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "pi_3PVAsqCuXDMKrYZe1FwwRbez",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pi_3PVAsqCuXDMKrYZe1FwwRbez",
    "payment_link": null,
    "profile_id": "pro_rhbVuxclv1iy3M2pZkkI",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_sKlZxW37AYQNFv4Z4LGy",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-06-24T11:31:26.868Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "127.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-06-24T11:16:29.856Z",
    "charges": null,
    "frm_metadata": null
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/hyperswitch_interfaces/src/integrity.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+pub trait FlowType<Request, T> {
+    /// Function to check to initiate integrity check
+    fn check_integrity(
+        &self,
+        request: &Request,
+        connector_transaction_id: Option<String>,
+    ) -> Result<(), IntegrityCheckError>;
+}
+
+impl<T, Request> FlowType<Request, T> for api::Authorize
```
Comment (jarnura) [inline crates/hyperswitch_interfaces/src/integrity.rs:39]: Implement this trait for flow specific router data.

File: crates/hyperswitch_interfaces/src/integrity.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+/// Trait to get connector integrity object based on request
+pub trait RequestIntegrity<T: ConnectorIntegrity> {
+    /// function to get response integrity object
+    fn get_response_integrity_object(&self) -> Option<T::IntegrityObject>;
+    /// function to get request integrity object
+    fn get_request_integrity_object(&self) -> T::IntegrityObject;
+}
+
+/// Trait to check flow type, based on which various integrity checks will be performed
+pub trait FlowType<Request, T> {
```
Comment (jarnura) [inline crates/hyperswitch_interfaces/src/integrity.rs:30]: add get_connector_id  function for this trait

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/hyperswitch_interfaces/src/integrity.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    type IntegrityObject;
+    /// helps in connector integrity check
+    fn compare(
+        req_integrity_object: Self::IntegrityObject,
+        res_integrity_object: Self::IntegrityObject,
+        connector_transaction_id: Option<String>,
+    ) -> Result<(), IntegrityCheckError>;
+}
+
+/// Trait to get connector integrity object based on request
```
Comment (jarnura) [inline crates/hyperswitch_interfaces/src/integrity.rs:18]:
```suggestion
/// Trait to get connector integrity object based on request and response
```

File: crates/hyperswitch_interfaces/src/integrity.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
+use common_utils::errors::IntegrityCheckError;
+use hyperswitch_domain_models::router_request_types::{
+    AuthoriseIntegrityObject, PaymentsAuthorizeData, PaymentsSyncData, SyncIntegrityObject,
+};
+
+/// Connector Integrity trait to check connector data integrity
+pub trait ConnectorIntegrity {
```
Comment (jarnura) [inline crates/hyperswitch_interfaces/src/integrity.rs:7]:
```suggestion
pub trait FlowIntegrity {
```

File: crates/hyperswitch_interfaces/src/integrity.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    /// helps in connector integrity check
+    fn compare(
+        req_integrity_object: Self::IntegrityObject,
+        res_integrity_object: Self::IntegrityObject,
+        connector_transaction_id: Option<String>,
+    ) -> Result<(), IntegrityCheckError>;
+}
+
+/// Trait to get connector integrity object based on request
+pub trait RequestIntegrity<T: ConnectorIntegrity> {
```
Comment (jarnura) [inline crates/hyperswitch_interfaces/src/integrity.rs:19]:
```suggestion
pub trait GetIntegrityObject<T: ConnectorIntegrity> {
```

File: crates/hyperswitch_interfaces/src/integrity.rs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    /// function to get response integrity object
+    fn get_response_integrity_object(&self) -> Option<T::IntegrityObject>;
+    /// function to get request integrity object
+    fn get_request_integrity_object(&self) -> T::IntegrityObject;
+}
+
+/// Trait to check flow type, based on which various integrity checks will be performed
+pub trait RouterDataType<Request, T> {
```
Comment (jarnura) [inline crates/hyperswitch_interfaces/src/integrity.rs:27]:
```suggestion
pub trait CheckIntegrity<Request, T> {
```

File: crates/router/src/core/payments/operations/payment_response.rs [hunk 7]
```diff
@@ -841,212 +841,251 @@ async fn payment_response_update_tracker<F: Clone, T: types::Capturable>(
...
-                } => {
-                    let connector_transaction_id = match pre_processing_id.to_owned() {
-                        types::PreprocessingResponseId::PreProcessingId(_) => None,
-                        types::PreprocessingResponseId::ConnectorTransactionId(
-                            connector_txn_id,
-                        ) => Some(connector_txn_id),
+            // match on connector integrity check
+            match router_data.integrity_check.clone() {
+                Err(err) => {
+                    println!("integrity flow check ");
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_response.rs:849]: remove

**Review by jarnura — COMMENTED** [review]
Add metrics for failed cases.

**Review by jarnura — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]
Dashboard specific changes looks fine.

**Pull request merged**